src/rust/runtime/network/transport.rs

```diff
@@ -0,0 +1,84 @@
+/// Transport API. This represents a higher level network API with full socket functionality.
+use crate::{
+    demikernel::config::Config,
+    runtime::{
+        fail::Fail,
+        memory::DemiBuffer,
+        scheduler::Yielder,
+        SharedDemiRuntime,
+    },
+};
+use ::socket2::{
+    Domain,
+    Type,
+};
+use ::std::{
+    fmt::Debug,
+    net::SocketAddr,
+};
+
+//======================================================================================================================
+// Trait Definition
+//======================================================================================================================
+
+/// This trait represents a high-level network API that supports both connection-based and connection-less
+/// communication using sockets.
+pub trait NetworkTransport: Clone + 'static {
+    type SocketDescriptor: Debug;
+
+    /// Create a new network transport.
+    fn new(config: &Config, runtime: &mut SharedDemiRuntime) -> Self;
+
+    /// Create a socket using the network transport layer.
+    fn socket(&mut self, domain: Domain, typ: Type) -> Result<Self::SocketDescriptor, Fail>;
+
+    /// Bind an address to the socket.
+    fn bind(&mut self, sd: &mut Self::SocketDescriptor, local: SocketAddr) -> Result<(), Fail>;
+
+    /// Listen on this socket in the network transport layer.
+    fn listen(&mut self, sd: &mut Self::SocketDescriptor, backlog: usize) -> Result<(), Fail>;
+
+    /// Forcibly close this socket in the network transport layer. This function should only be used in Drop and other
+    /// internal functions, never exposed to the application.
```

Comment:
Can this (not to be exposted outside) be controlled/restricted at compile-time? Maybe be structuring differently (or hierarchy etc.)?

Reply:
Yes, I believe  since it is not exposed through the demikernel libos layer, then external apps cannot use it. I don't know if we can enforce that we do not expose it in some other way.